hashintel/hash
# Implement `Termination` for `error_stack::Report`

## 🌟 What is the purpose of this PR?

This adds custom error code support for `error-stack`

## 🔗 Related links

Closed #670.

Though the implementation has changed since I wrote the issue and learned more about the code base.

## 🔍 What does this change?

- ~~Adds a method `attach_exit_code(mut self, exit_code: u8) -> Self` on the `Report<C>` struct.~~
- ~~Adds a new field `exit_code: Option<u8>` on the `ReportImpl` struct.~~
- Implements the `Termination` trait for the `Report<C>` struct.

## 📜 Does this require a change to the docs?

The docs have been added.

## 🐾 Next steps

Please let me know if there are any changes that should be made.

## 🛡 What tests cover this?

There is the test doc test that passes. Let me know if I should add a separate test.

## ❓ How to test this?

1.  Pull the forked repo.
2. Check out the `feature/custom-exit-code` branch.
3. Run `cargo test --doc`

## 📹 Demo

![image](https://user-images.githubusercontent.com/16431709/174541437-f432c572-b876-44e3-94dd-55bc4b9115a5.png)
